File: Data Structures & Algorithms/surrounded-regions/submission-4.java
class Solution {
    private int[] DR = {-1,0,1,0};
    private int[] DC = {0,-1,0,1};

    public void solve(char[][] board) {
        int m=board.length, n=board[0].length;
        boolean[][] visited = new boolean[m][n];
        for(int r=0; r<m; r++){
            for(int c=0; c<n; c++){
                if(board[r][c] == 'X'){
                    visited[r][c] = true;
                }
                if(r == 0 || r == m-1 || c == 0 || c == n-1){
                    if(board[r][c] == 'O')
                        dfs(board, r, c, visited);
                }
            }
        }
        for(int r=0; r<m; r++){
            for(int c=0; c<n; c++){
                if(!visited[r][c]){
                    board[r][c] = 'X';
                }
            }
        }
    }
    
    private void dfs(char[][] board, int r, int c, boolean[][] vis){
        int m = board.length, n = board[0].length;
        if(vis[r][c] || board[r][c] == 'X') return;
        vis[r][c] = true;

        for(int i=0; i<4; i++){
            int nr = r + DR[i];
            int nc = c + DC[i];
            if(nr >= 0 && nr < m && nc >= 0 && nc < n){
                dfs(board, nr, nc ,vis);
            }
        }
    }
}

File: Data Structures & Algorithms/count-paths/submission-4.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[1] * (n) for _ in range(m)]
        for i in range(m-2, -1, -1):
            for j in range(n-2, -1, -1):
                dp[i][j] = dp[i+1][j] + dp[i][j+1]
        return dp[0][0]
        dp = {}
        def dfs(i, j):
            if i < 0 or i >= m or j < 0 or j >= n:
                return 0
            if i == m-1 and j == n-1:
                return 1
            if (i, j) in dp:
                return dp[(i, j)]
            
            dp[(i, j)] = dfs(i, j+1) + dfs(i+1, j)
            return dp[(i, j)]

        return dfs(0,0)